File: src/pratice/binarysearch/용돈관리6236.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class 용돈관리6236 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, M;
    public static int[] A;

    public static void main(String[] args) throws IOException {
        input();
        pro();
    }

    private static void pro() {
        int L = 0;
        int R = 10_000;
        int k = R;
        while(L <= R){
            int mid = (L + R) / 2;
            if(isPossible(mid)){
                k = mid;
                R = mid - 1;
            }else{
                L = mid + 1;
            }
        }
        System.out.println(k);
    }

    private static boolean isPossible(int mid) {
        int cnt = 1;
        long sum = 0;
        for (int i = 0; i < N; i++) {
            if(sum + A[i] <= mid){
                sum += A[i];
            }else{
                sum = A[i];
                cnt++;
            }
        }
        return cnt >= M;
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        M = Integer.parseInt(split[1]);
        A = new int[N];
        split = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(split[i]);
        }
    }
}

File: src/pratice/binarysearch/용돈 관리6236.kt
package pratice.binarysearch

import kotlin.math.max

fun main() {
    val (N, M) = readln().split(" ").map { it.toInt() }
    val arr = mutableListOf<Int>()

    repeat(N) {
        arr.add(readln().toInt())
    }

    lowerBound(arr, M)
}

private fun lowerBound(arr: MutableList<Int>, m: Int) {
    var L = arr[0]
    var R = 1000000000
    for (num in arr) L = max(L, num)
    var res = R
    while (L <= R) {
        val mid = (L + R) / 2
        if (determination(arr, m, mid)) {
            res = mid
            R = mid - 1
        } else {
            L = mid + 1
        }
    }
    println(res)
}

private fun determination(arr: MutableList<Int>, m: Int, mid: Int): Boolean {
    var sum = 0
    var cnt = 0

    for (num in arr) {
        if (sum + num <= mid) {
            sum += num
        } else {
            cnt ++
            sum = num
        }
    }
    return cnt < m
}
